wasm/derive/src/smartcontract.rs

```diff
@@ -0,0 +1,60 @@
+use super::*;
+
+pub fn impl_entrypoint(_: TokenStream, item: TokenStream) -> TokenStream {
+    let syn::ItemFn {
+        attrs,
+        vis,
+        sig,
+        mut block,
+    } = parse_macro_input!(item);
+
+    verify_function_signature(&sig);
+    let fn_name = &sig.ident;
+
+    block.stmts.insert(
+        0,
+        parse_quote!(
+            use iroha_wasm::Execute as _;
+        ),
+    );
+
+    quote! {
+        // NOTE: The size of the `len` parameter is defined by the target architecture
+        // which is `wasm32-unknown-unknown` and therefore not dependent by the architecture
+        // smart contract is compiled on or the architecture smart contract is run on
+        /// Smart contract entry point
+        ///
+        /// # Safety
+        ///
+        /// Given pointer and length must comprise a valid memory slice
+        #[no_mangle]
+        pub unsafe extern "C" fn _iroha_wasm_main(ptr: *const u8, len: usize) {
+            #fn_name(
+                iroha_wasm::_decode_from_raw::<
+                    <iroha_wasm::data_model::account::Account as iroha_wasm::data_model::Identifiable>::Id>
+                (ptr, len)
+            )
+        }
+
+        #[allow(clippy::needless_pass_by_value)]
```

Comment:
This doesn't look right.

Reply:
Fixed. I just moved that from the old code

Reply:
From the smartcontract writer point of view it makes sense to pass parameters by values. It may use them or may not